iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp

```diff
@@ -133,11 +133,207 @@ LogicalResult clearStruct(OpBuilder builder, Value structValue,
+
+      return None;
+    }
+    ordinal = op->getAttr(attrName)
+                  .cast<ArrayAttr>()[index]
+                  .cast<IntegerAttr>()
+                  .getInt();
+  } else {
+    // refValue is a function argument.
+    op = parentFuncOp.getOperation();
+    unsigned index;
+    for (BlockArgument blockArg : parentFuncOp.getArguments()) {
+      if (blockArg == refResult) {
+        index = blockArg.getArgNumber();
+      }
+    }
+    if (!op->hasAttr("argument_ordinal")) {
+      parentFuncOp.emitError() << "missing annotation for ordinal";
+
+      return None;
+    }
+    ordinal = op->getAttr("argument_ordinal")
+                  .cast<ArrayAttr>()[index]
+                  .cast<IntegerAttr>()
+                  .getInt();
+  }
+
+  return findRefByOrdinal(builder, location, parentFuncOp, ordinal);
+}
+
+LogicalResult annotateFuncOp(IREE::VM::FuncOp &vmFuncOp, mlir::FuncOp &funcOp,
+                             RegisterAllocation &registerAllocation,
+                             ValueLiveness &valueLiveness) {
+  auto ctx = funcOp.getContext();
+
+  OpBuilder builder(funcOp);
+
+  funcOp.getOperation()->setAttr("emitc.static", UnitAttr::get(ctx));
+
+  // Annotate new function with a calling convention string which gets used in
+  // the CModuleTarget.
+  funcOp.getOperation()->setAttr(
+      "vm.calling_convention",
+      StringAttr::get(ctx, makeCallingConventionString(vmFuncOp).getValue()));
+
+  auto isRef = [](Value value) {
+    return value.getType().isa<IREE::VM::RefType>();
+  };
+
+  auto getOrdinal = [&isRef, &registerAllocation](Value value) -> int64_t {
+    if (isRef(value)) {
+      return registerAllocation.mapToRegister(value).ordinal();
+    }
+    // Return a default value which never gets used.
+    return 0;
+  };
+  auto isLastUse = [&isRef, &valueLiveness](Value value,
+                                            Operation &op) -> bool {
+    if (isRef(value)) {
+      return valueLiveness.isLastValueUse(value, &op);
+    }
+    // Return a default value which never gets used.
+    return false;
+  };
+
+  SmallVector<int64_t> ordinals =
+      llvm::to_vector<1>(llvm::map_range(funcOp.getArguments(), getOrdinal));
+  funcOp.getOperation()->setAttr("argument_ordinal",
+                                 builder.getIndexArrayAttr(ordinals));
+
+  for (Block &block : funcOp.getBlocks()) {
+    // Annotate operations with vm analysis
+    for (Operation &op : block.getOperations()) {
+      if (llvm::any_of(op.getOperands(), isRef)) {
+        SmallVector<int64_t> ordinals =
+            llvm::to_vector<1>(llvm::map_range(op.getOperands(), getOrdinal));
+        SmallVector<bool> lastUse = llvm::to_vector<1>(llvm::map_range(
+            op.getOperands(),
+            [&op, &isLastUse](auto v) { return isLastUse(v, op); }));
+
+        op.setAttr("operand_ordinal", builder.getIndexArrayAttr(ordinals));
+        op.setAttr("operand_is_last_use", builder.getBoolArrayAttr(lastUse));
```

Comment:
just a heads up that this approach is going to be very very slow - I don't recall why these started being in the IR instead of in an analysis? each attribute mutation on an op causes an entirely new attribute dictionary to be created, uniqued, and stored forever. It's fine if it really is something structural in the IR, but for things like this analysis that are easy to recompute from the IR source of truth this is not a the best approach. The information being added should also not be so volatile as this: adding things to the IR should be done when they are semi-stable such that going in and hand transposing two ops (or having a pattern that does the same) which is structurally safe (think transposing the order of two constant ops) shouldn't break the program, but here it would.

Reply:
I tried to keep a hash map from `iree_vm_ref_t*` to `IREE::VM::Ref` so I could look up the original Value and then query the analysis for it during the conversion. That seemed to work until I started removing the ref basic block arguments. I think that invalidated the hash map values because the block arguments were destroyed.